src/test/java/states/InvalidOpCodeStateTest.java

```diff
@@ -0,0 +1,185 @@
+
+        } catch (IOException e) {
+            Assert.fail(e.getMessage());
+        }
+    }
+
+    @Test
+    public void testInvalidData() {
+        try {
+            checker = new ErrorChecker(Message.MessageType.DATA, 1);
+            state.setErrorChecker(checker);
+            byte[] expectedDataBytes = new DataMessage(1, new byte[] { 0 }).toByteArray();
+            invalidOPData = duplicateDataInvOp(expectedDataBytes);
+            DatagramPacket expectedPacket = new DatagramPacket(expectedDataBytes, expectedDataBytes.length, serverSocketAddress);
+            DatagramPacket invalidPacket = new DatagramPacket(invalidOPData, invalidOPData.length, serverSocketAddress);
+            Mockito.when(socket.receive())
+                    .thenReturn(expectedPacket)
+                    .thenThrow(new RuntimeException("TEST EXCEPTION"));
+
+            state.setServerWorkerPort(3000);
+            state.setClientAddress(serverSocketAddress);
+            state.execute();
+
+            InvalidPacketMatcher matcher = new InvalidPacketMatcher(invalidPacket);
+            Mockito.verify(socket, Mockito.times(0)).forwardPacket(Mockito.eq(expectedPacket), Mockito.eq(serverAddress), Mockito.eq(3000));
+            Mockito.verify(socket, Mockito.times(1)).forwardPacket(Mockito.argThat(matcher), Mockito.eq(serverAddress), Mockito.eq(3000));
+        } catch (IOException e) {
+            Assert.fail(e.getMessage());
+        }
+    }
+
+    @Test
+    public void testInvalidACK() {
+        try {
+            checker = new ErrorChecker(Message.MessageType.ACK, 1);
+            state.setErrorChecker(checker);
+
+            byte[] expectedACKBytes = new AckMessage(1).toByteArray();
+            invalidOPData = duplicateDataInvOp(expectedACKBytes);
+            
+            DatagramPacket expectedPacket = new DatagramPacket(expectedACKBytes, expectedACKBytes.length, serverSocketAddress);
+            DatagramPacket invalidPacket = new DatagramPacket(invalidOPData, invalidOPData.length, serverSocketAddress);
+            Mockito.when(socket.receive())
+                    .thenReturn(expectedPacket)
+                    .thenThrow(new RuntimeException("TEST EXCEPTION"));
+
+            state.setServerWorkerPort(3000);
+            state.setClientAddress(serverSocketAddress);
+            state.execute();
+
+            InvalidPacketMatcher matcher = new InvalidPacketMatcher(invalidPacket);
+            Mockito.verify(socket, Mockito.times(0)).forwardPacket(Mockito.eq(expectedPacket), Mockito.eq(serverAddress), Mockito.eq(3000));
+            Mockito.verify(socket, Mockito.times(1)).forwardPacket(Mockito.argThat(matcher), Mockito.eq(serverAddress), Mockito.eq(3000));
+        } catch (IOException e) {
+            Assert.fail(e.getMessage());
+        }
+    }
+    
+    public byte[] duplicateDataInvOp(byte[] arr) {
+    	byte [] arrVal = new byte[arr.length];
```

Comment:
You can use [`Array.copyOf(arr, arr.length)`](https://docs.oracle.com/javase/7/docs/api/java/util/Arrays.html#copyOf(byte[],%20int))